src/optimizer/build_probe_side_optimizer.cpp

```diff
@@ -64,6 +63,44 @@ static inline idx_t ComputeOverlappingBindings(const vector<ColumnBinding> &hays
 	return result;
 }
 
+BuildSize BuildProbeSideOptimizer::GetBuildSizes(LogicalOperator &op) {
+	BuildSize ret;
+	switch (op.type) {
+	case LogicalOperatorType::LOGICAL_COMPARISON_JOIN:
+	case LogicalOperatorType::LOGICAL_CROSS_PRODUCT:
+	case LogicalOperatorType::LOGICAL_ANY_JOIN:
+	case LogicalOperatorType::LOGICAL_DELIM_JOIN: {
+		auto &left_child = op.children[0];
+		auto &right_child = op.children[1];
+
+		auto left_column_count = left_child->GetColumnBindings();
+		auto right_column_count = right_child->GetColumnBindings();
+
+		// resolve operator types to determine how big the build side is going to be
+		op.ResolveOperatorTypes();
+		auto left_column_types = left_child->types;
+		auto right_column_types = right_child->types;
+
+		idx_t left_build_side = 0;
+		for (auto &type : left_column_types) {
+			left_build_side += GetTypeIdSize(type.InternalType());
```

Comment:
Can we use the type visitor to recurse into nested types here? (`duckdb/common/type_visitor.hpp`)

Reply:
Maybe we also want to add a constant per column (because likely 8 tinyint columns are more expensive to handle than one bigint column) - also because of the nullmask and such. Maybe experiment with this as well?

Reply:
We've solved these two issues by using the `TupleDataLayout` to get the actual row width (including nullmask), and adding a penalty per column, so that we not only take the total width but also the cost of materializing each column into account. This is good to go from my side now